app/services/add_attachment_service.rb

```diff
@@ -0,0 +1,63 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddAttachmentService
+  attr_reader :container, :author
+
+  def initialize(container, author:)
+    @container = container
+    @author = author
+  end
+
+  ##
+  # Adds and saves the uploaded file as attachment of the given container.
+  # In case the container supports it, a journal will be written.
+  #
+  # An ActiveRecord::RecordInvalid error is raised if any record can't be saved.
+  def add_attachment(uploaded_file:, description:)
+    attachment = Attachment.new(file: uploaded_file,
+                                container: container,
+                                description: description,
+                                author: author)
+    save attachment
```

Comment:
I agree with the reusability of the service and my proposal does not need to pass `uploaded_file` and `description` into the service constructor. 

The service definitely shouldn't return two different types of objects depending on the operation success. What I use in all services and have seen consistently is a boolean result of the service operation.

In the end, this is the exact identical pattern to ActiveRecord's save. You get a boolean result that you can continue in the best case, and get an accessor to errors in the worst case. This comes down to a matter of taste and I only brought that up due to the way most of the other services in OP are built.